Better Error Handling

bstee615 opened the pull request:
Refactored calls to methods to use Promise error handling rather than try/catch. Promise chains are new to me, but I think it makes the code in the api/ modules cleaner because it eliminates try/catch from those places. Also, some places got auto-formatted by VS Code. Sorry I didn't update you on my plans before this req. Cheers!

Commit:
Changing some things.

Commit:
Changing some things.

Commit:
fix(*): refactorings for error handling quality

Change getRepositories() to fix the bug you were experiencing - early,
silent quits on user not found. Refactor methods to come along with
Promise-chain error handling.

Resolves #19

Commit:
chore(app.js): remove console logs I used for debugging and add comments

Want to keep things neat!

jwu910 reviewed (changes requested):
@bstee615 Thanks for shooting this PR over! First of all, good to see that this first iteration of the project was easily enough understood.

I'd just like to clarify a few things regarding the issue requirements. I was hoping to keep the use of try/catch in conjunction with my async/await calls. I had one giant try/catch but I was hoping to implement smaller try/catch blocks around more of the granular await lines so we can see which promises are throwing errors.

I feel that If we move to chaining promises, we will easily fall into callback hell which can make the code a little harder to follow imo. 

Would you mind taking a shot at this using that approach? Here's a little more info on [try/catch vs promise chains](https://javascript.info/async-await) Let me know if you have questions, i'd be happy to work on this with you

jwu910 commented:
Also regarding the auto formatting, Your editor should be picking up the .prettierrc and .editorconfig. I'll take a look to see if i have those set up correctly. You shouldn't need to worry about formatting for this project. There should also be a pre-commit hook that re-formats your code on commit, but if thats not working I need to fix that too..

jwu910 commented:
#19

bstee615 commented:
Yes, I will convert to the try/catch approach. I agree that it's too much hassle to add extra links in the chain; frankly, I'm new to NodeJS so I am not sure what's idiomatic yet. Any code style/refactorings should be up to your discretion, since you will be the one maintaining it :)

I think my approach will be to split the Promise chain up into multiple methods, each with their try/catch block inside, In my mind this will be cleaner than the alternative. Will update over the next few days.

jwu910 commented:
@bstee615 Right on, yeah most my experience has been working with promise chains, trying to clean that up with async/await.

So far the patterns I've seen have been something like this
```
try {
  await response = somePromise();
} catch (error) {
  // report error related to promise above
}
```

Commit:
refactor(app.js):switch to try/catch over Promise chains

try/catch is more scalable and idiomatic.

Addresses review on #20

bstee615 commented:
Yep, just got them added. Ended up being a pretty simple refactor once I figured out that **var** scopes into the nearest function block. That makes it a lot easier to cleanly establish state between multiple try blocks. I pushed a commit for review.

jwu910 commented on src/api/cloneRepo.js:
```diff
@@ -37,5 +37,10 @@ const execGit = args => {
+    .then(data => data)
+    .catch(function(errmsg) {
+      throw new Error(errmsg);
+    });
```
Hey @bstee615 couple preferences with this. This function returns a promise, so we're alright not using a `.then` chain here. Cause we'll be calling this function inside an async func in the main application

jwu910 commented on src/api/forkRepo.js:
```diff
@@ -3,23 +3,18 @@ import axios from 'axios';
 import auth from '../auth';
 
 export const forkRepo = async ({ repository: { owner, repo } }) => {
```
if we're not using `await` in this function, i think we can scrap the `async` here

jwu910 commented on src/app.js:
```diff
@@ -8,15 +7,48 @@ import { cloneRepo } from './api/cloneRepo';
-const notifier = updateNotifier({ pkg });
+const notifier = updateNotifier({
+  pkg,
+});
```
This is an unnecessary formatting change, did the auto formatter not adjust this? I was trying to set up pretty-quick to auto format commits through the commitizen prompts. If it didnt work, lmk--I may need to check my project configs

jwu910 commented on src/app.js:
```diff
@@ -8,15 +7,48 @@ import { cloneRepo } from './api/cloneRepo';
 import { forkRepo } from './api/forkRepo';
 import { getRepos } from './api/getRepos';
 import { truncate } from './util/truncate';
+import { exists } from 'fs';
```
Are we using this library? I think we may have an unnecessary import

jwu910 commented on src/app.js:
```diff
@@ -25,11 +57,9 @@ export const start = async args => {
 
-  let reposList;
-
-  //TODO: prompt for username if none provided
```
We haven't completed this TODO, we should leave it in. It's not in the scope of this issue

jwu910 replied:
Removing this comment wasn't in the scope of this ticket, let's leave it in

jwu910 commented on src/app.js:
```diff
@@ -38,61 +68,47 @@ export const start = async args => {
-    const response = await getRepos(username);
-
-    const repositories = response.user.repositories.edges;
+    var userGraph = await getRepos(username);
```
We should be using `const` here. Technically runs, but `var` working with ES6, and the introduction of `let/const` we have more explicit ways to declare variables without `var`.

bstee615 replied:
Ok thanks, I wasn’t aware of that.

jwu910 replied:
Yup! If you look up ‘Const vs let vs var’ You’ll find a bunch of good reads

jwu910 commented on src/app.js:
```diff
@@ -8,15 +7,48 @@ import { cloneRepo } from './api/cloneRepo';
 }
 
-export const start = async args => {
+// Extracts repository info from GraphQL edge nodes
```
Let's remove all non-todo comments from committed code

jwu910 commented on src/app.js:
```diff
@@ -38,61 +68,47 @@ export const start = async args => {
+    var userGraph = await getRepos(username);
+  } catch (err) {
+    // This error message from the GitHub API is especially ugly.
+    process.stderr.write(`Error: could not get user ${username}.\n`);
```
Let's state in our own message where this error was thrown and then spit out the actual error message because those error logs are still helpful.

```
process.stderr.write(`Error: Could not get repos from user: ${username}.\n`;
process.stderr.write(error);
```

jwu910 commented on src/app.js:
```diff
@@ -38,61 +68,47 @@ export const start = async args => {
+    // Check user & repo
+    const user = userGraph.user;
+    if (!user) {
+      throw new Error('user not found.');
```
Sample response from the github api explorer

```
{
  "data": {
    "user": null
  },
  "errors": [
    {
      "type": "NOT_FOUND",
      "path": [
        "user"
      ],
      "locations": [
        {
          "line": 2,
          "column": 9
        }
      ],
      "message": "Could not resolve to a User with the login of 'alskhe2'."
    }
  ]
}
```
Can we just grab the errors `message` and `type` as part of what we report back?

jwu910 reviewed (changes requested):
@bstee615 	Few changes, please see my comments.

Let me know if you have questions. I will take a look into the formatting issue soon.

It also looks like you had some accidental commits at the beginning of your PR. net zero commits like that in the future could probably just be squashed or dropped so they don't show up in history. :+1: 

Thanks again for the PR, looking forward to reviewing the changes.

bstee615 commented:
@jwu910 Sounds great, thanks for the thorough review. I learned some new things. Should i go in manually and re-commit the changes you requested, or will github emplace those changes after I resolve them?

Commit:
refactor getReposList into a separate function

Commit:
move error handling to grab user-friendly error message from getRepos()

bstee615 reviewed (commented):
I submitted two commits implementing these changes. It was a good call, your suggestions removed all occurrences of `var` and shortened `app.js` by 20 lines.

bstee615 commented:
@jwu910 Not sure how to request another review, but the previous two comments should cover all your requested changes. Can you please take a look over and, if you're content, close this PR?

jwu910 commented on src/api/getRepos.js:
```diff
@@ -1,32 +1,30 @@
+  try {
+    return await client.request(query);
+  } catch (e) {
+    throw new Error(e.response.errors[0].message);
```
Lets follow the pattern that we've established here breaking down the try-catches into more modular blocks in app.js.

Lets make this function not async and just return the promise `client.request(query)`, also lets be explicit with returning (error) rather than just (e) so it's more explicit

bstee615 replied:
`graphql-request` library throws an error with the following large chunk of data. I think the only thing we should include from this data is the message. I could just return `client.request(query)` and remove the error'handling code from this function if we added better try/catch blocks in `start()`.

```
Error: Could not resolve to a User with the login of 'jwu911'.: {"response":{"data":{"user":null},"errors":[{"type":"NOT_FOUND","path":["user"],"locations":[{"line":2,"column":5}],"message":"Could not resolve to a User with the login of 'jwu911'."}],"status":200},"request":{"query":"{\n    user(login: \"jwu911\") {\n      name\n      repositories(first:100 orderBy:{field: STARGAZERS, direction: DESC}) {\n        edges {\n          node {\n            id\n            name\n            description\n            forks {\n              totalCount\n            }\n            stargazers {\n              totalCount\n            }\n          }\n        }\n      }\n    }\n  }"}}
```

jwu910 replied:
Yeah, so i think these functions should return their promises.

I haven't yet tested every use case for these errors, so I think it'll be good to either explore the error messages returned in the different try/catch blocks in our `start()` function and if the shape of the response is different for each error, we can log 2 messages

1. We've hit in error doing [X]
2. Message: [return error message]

so if we're getting repos, i can image an error to be thrown like:
```
Error fetching repositories.
Message: //some long error string
```

I think we can have a longer hard-to-read error string after if we provide first a shorter error message summarizing what happened

jwu910 commented on src/app.js:
```diff
@@ -16,7 +15,36 @@ if (notifier.update) {
 }
 
-export const start = async args => {
+function reposListFromGraphEdges(edges, username) {
```
I dont think we need to abstract the functionality this function is providing. We're building the prompt that we're using which I would consider a core functionality of the main application.

This may make the `start()` func a little longer, but i think for the state of this app, it makes sense. If the prompts get too complex, i think we can then abstract it out to utility functions similar to the `getRepos()` and `forkRepo()` utils we have set up here.

bstee615 replied:
I agree - I am just used to writing very short functions as my own style, but I want to comply with the precedents you've set.

jwu910 commented on src/app.js:
```diff
@@ -25,74 +53,38 @@ export const start = async args => {
+    );
 
-    let selection = await prompts({
+    const chosenRepo = await prompts({
```
`let` -> `const` is a great catch here, however I think the variable name is an unnecessary change.

jwu910 commented on src/app.js:
```diff
@@ -25,74 +53,38 @@ export const start = async args => {
-    );
+    if (!chosenRepo.repository) {
+      throw new Error('repository not found.');
+    }
```
repository is the return object from the user prompt, the only time this object comes back falsy would be if a user killed the process early such as via `ctrl+c`

bstee615 replied:
So, for this one, `repository` does return null if the user searches for a repo name that does not exist. To show it, I added a line to print out the value of selection after calling `prompts()`.
```
C:\code\oss\git-started>node index.js -u jwu910
√ Select jwu910's repository to fork or start typing to search » adsd

{ repository: null } <- Actual value of selection

...
```

jwu910 replied:
Oh! Right on. I see what you did here, I was thinking this was catching empty repositories if the user entered an invalid github user name.

jwu910 commented on src/app.js:
```diff
@@ -25,74 +53,38 @@ export const start = async args => {
+    await cloneRepo(program.ssh ? fork.data.ssh_url : fork.data.clone_url);
+    process.stdout.write(`Successfully forked and cloned repo.\n`);
+  } catch (err) {
+    process.stderr.write(`${err}\n`);
```
Couple comments about these lines

When I mentioned the modular try/catch blocks i had envisioned something like this:

```
try {
	const fork = await forkRepo(selection);

	console.log('Successfully forked repo');
}
catch (error) {
	console.error('Unable to fork repo');
	console.error('Reason: ', error);

	process.exit(1);
}

try {
	await cloneRepo(program.ssh ? fork.data.ssh_url : fork.data.clone_url);

	console.log('Successfully cloned repository');
}
catch (error) {
	console.error('Unable to clone repo');
	console.error('Reason: ', error);

	process.exit(1);
}

```

I'd have to see this snippet in a larger scope to see what the formatting looks like impelmenting the try/catches, but from my readings about the use of async/await over promise.then()'s uses modular try catches similar like this.

I'll try to do some research in the mean time, but let me know what you think about this

bstee615 replied:
This code actually doesn't run. `const` variables are block-scoped, so any code outside the try/catch block can't access them. That's why I had added a `var` earlier; `var`-declared variables are function-scoped, so they can be seen outside the try/catch where they are declared. AFAIK, there are two allternative ways of doing it:

One is what I had tried before, where I had split each try/catch into its own function. This way, I can return the primary "output" of the try/catch block.

Another way that I avoided goes like this:
```
let fork;
try {
    fork = await forkRepo();
    ...
}
```
This way, `fork` is explicitly scoped outside of the try/catch block. This seems like a more "ES6" way to do it. I didn't do it because I just don't like the way it looks. You can make the call, really, on what's the right way to do it.

jwu910 replied:
your comment about `let` is accurate. 

And we'll use that. You can read up more about var/const/let, but the most common patterns I've seen now is to avoid `var` all together with the introduction of `let/const`. 

But this comment was more focused on the structure of the `try/catches`. And this was just typed up as an example.

https://github.com/jwu910/git-started/pull/20/files/9c41ffa858ebe0a78b36176ea0def1bbfda74db3#r261891765 I made a comment further up about why I think separate try catches would be the way to go.

bstee615 replied:
Okay, I’ll go with *let*. That will let me restructure the try/catch blocks better.

jwu910 commented on src/api/cloneRepo.js:
```diff
@@ -36,6 +36,5 @@ const execGit = args => {
 
 export const cloneRepo = url => {
   const args = ['clone', url, '-v'];
-
```
Unnecessary line removal

jwu910 commented on src/app.js:
```diff
@@ -25,74 +53,38 @@ export const start = async args => {
-      };
-    });
+    const userGraph = await getRepos(username);
+    const reposList = reposListFromGraphEdges(
```
See my comment above about the reposListFromGraphEdges abstraction.

Note about the line spacing here too. I would like to keep like declarations together, but line break between declarations that are dependent on the previous line. 

example:

```
const myFoo = 'foo';
const myBar = 'bar';

// and
const myFoo = 'foo';

const myBar = someFunc(myFoo);
```

jwu910 reviewed (changes requested):
@bstee615 Thanks again for the hard work! I appreciate it!

I've requested a few more changes regarding the approach we're taking to this ticket, please let me know if you have questions.

Also, I've pushed some changes up to development, so please pull the latest updates in. You should be able to make your commits using the `npm run commit` script or you can install commitizen and take a look at their conventions. I don't have a commit message linter set up yet, but will be set up soon that will check for these commit message formats. Thanks!

Commit:
Merge branch 'development' of https://github.com/jwu910/git-started into user-validation

Commit:
refactor(app.js): implement some requested code review changes

relocate core app logic to start(), minor style fixes, un-delete "TODO prompt for username" comment

re #19

jwu910 reviewed (changes requested):
Few more notes about the formatting. Please leave a line break between non-like lines. We're not too worried about the number of lines for this CLI, if we start hitting a really large set of lines, we may want to look at abstracting the code in to modular utils at that point, but it seems fine for now

When you run the commit command, are you getting the prompt that your code's being formatted as well?

Commit:
refactor(app.js, getRepos.js): reshuffle error messages for better readability/granularity

move to modular try/catch, remove try/catch in getRepos(), more detailed error messages.

ref #19

bstee615 commented:
@jwu910 Ok, I've pushed another commit with the changes you requested. I agree that it does feel cleaner and easier to extend.

After my latest commit, all the requirements of the original issue are filled: added error handling to validate user and repository fetching. We good to submit this one?

jwu910 commented:
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

jwu910 commented:
Thanks merged!

bstee615 commented:
@jwu910 This was fun. Maybe in the future I will try to help out with some other issues (and be more diligent about exploring possible solutions so that there's not so much back-and-forth). Thank you!

jwu910 commented:
@bstee615 no problem at all! Welcome to open source, every project's got their standards and patterns, you'll see plenty of code that looks better than this, you may see some code thats not--but it's all good, everyone's learning every day they code.

Thanks for taking the time to contribute to this
